src/Bicep.Core/Emit/TemplateWriter.cs

```diff
@@ -360,7 +365,8 @@ FullyQualifiedAmbientTypeReferenceExpression fullyQualifiedAmbientTypeReference
                         @bool.SourceSyntax),
                 },
                 @bool.SourceSyntax),
-            UnionTypeExpression unionType => GetTypePropertiesForUnionTypeExpression(unionType),
+            UnionTypeExpression unionType when TypeHelper.TryRemoveNullability(unionType.ExpressedUnionType) is not DiscriminatedObjectType
+                => GetTypePropertiesForUnionTypeExpression(unionType),
```

Comment:
With the last batch of commits, a nullable discriminated union should always be a NullableTypeExpression whose base expression is a DiscriminatedObjectTypeExpression, shouldn't it?

Reply:
Will look into this again. I think you're right.